File: backend/app/models/performance.py
from __future__ import annotations

import uuid
from datetime import date, datetime

from sqlalchemy import CheckConstraint, Date, DateTime, ForeignKey, Integer, String, Text, UniqueConstraint
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base
from app.models.base import TimestampMixin, UUIDPkMixin


class PerformanceCycle(UUIDPkMixin, TimestampMixin, Base):
    __tablename__ = "performance_cycles"
    __table_args__ = (
        CheckConstraint(
            "status IN ('draft', 'pending_approval', 'open', 'review', 'calibration', 'published', 'closed')",
            name="ck_performance_cycle_status",
        ),
    )

    name: Mapped[str] = mapped_column(String(150), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    start_date: Mapped[date] = mapped_column(Date, nullable=False)
    end_date: Mapped[date] = mapped_column(Date, nullable=False)
    goal_due_date: Mapped[date] = mapped_column(Date, nullable=False)
    self_review_due_date: Mapped[date] = mapped_column(Date, nullable=False)
    manager_review_due_date: Mapped[date] = mapped_column(Date, nullable=False)
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="draft", server_default="draft")
    created_by_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("employees.id"), nullable=False)
    approved_by_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("employees.id"))
    approved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    published_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    created_by: Mapped[Employee] = relationship("Employee", foreign_keys=[created_by_id])  # noqa: F821
    approved_by: Mapped[Employee | None] = relationship("Employee", foreign_keys=[approved_by_id])  # noqa: F821


class PerformanceGoal(UUIDPkMixin, TimestampMixin, Base):
    __tablename__ = "performance_goals"
    __table_args__ = (
        CheckConstraint("weight > 0 AND weight <= 100", name="ck_performance_goal_weight"),
        CheckConstraint("progress >= 0 AND progress <= 100", name="ck_performance_goal_progress"),
        CheckConstraint(
            "status IN ('draft', 'pending_approval', 'approved', 'changes_requested')",
            name="ck_performance_goal_status",
        ),
    )

    cycle_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("performance_cycles.id", ondelete="CASCADE"), index=True
    )
    employee_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("employees.id", ondelete="CASCADE"), index=True)
    title: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    category: Mapped[str] = mapped_column(String(40), nullable=False)
    measurement: Mapped[str] = mapped_column(Text, nullable=False)
    weight: Mapped[int] = mapped_column(Integer, nullable=False)
    target_date: Mapped[date] = mapped_column(Date, nullable=False)
    progress: Mapped[int] = mapped_column(Integer, nullable=False, default=0, server_default="0")
    evidence: Mapped[str | None] = mapped_column(Text)
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="draft", server_default="draft")
    manager_comment: Mapped[str | None] = mapped_column(Text)
    version: Mapped[int] = mapped_column(Integer, nullable=False, default=1, server_default="1")

    cycle: Mapped[PerformanceCycle] = relationship()
    employee: Mapped[Employee] = relationship("Employee")  # noqa: F821


class PerformanceReview(UUIDPkMixin, TimestampMixin, Base):
    __tablename__ = "performance_reviews"
    __table_args__ = (
        UniqueConstraint("cycle_id", "employee_id", name="uq_performance_review_cycle_employee"),
        CheckConstraint(
            "status IN ('not_started', 'self_draft', 'submitted_to_manager', "
            "'manager_submitted', 'calibrated', 'published', 'acknowledged')",
            name="ck_performance_review_status",
        ),
    )

    cycle_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("performance_cycles.id", ondelete="CASCADE"), index=True
    )
    employee_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("employees.id", ondelete="CASCADE"), index=True)
    manager_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("employees.id"))
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="not_started", server_default="not_started")
    self_summary: Mapped[str | None] = mapped_column(Text)
    self_rating: Mapped[int | None] = mapped_column(Integer)
    manager_summary: Mapped[str | None] = mapped_column(Text)
    manager_rating: Mapped[int | None] = mapped_column(Integer)
    calibration_comment: Mapped[str | None] = mapped_column(Text)
    calibrated_rating: Mapped[int | None] = mapped_column(Integer)
    calibrated_by_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("employees.id"))
    final_rating: Mapped[int | None] = mapped_column(Integer)
    published_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    acknowledged_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    employee_comment: Mapped[str | None] = mapped_column(Text)

    cycle: Mapped[PerformanceCycle] = relationship()
    employee: Mapped[Employee] = relationship("Employee", foreign_keys=[employee_id])  # noqa: F821
    manager: Mapped[Employee | None] = relationship("Employee", foreign_keys=[manager_id])  # noqa: F821
    calibrated_by: Mapped[Employee | None] = relationship("Employee", foreign_keys=[calibrated_by_id])  # noqa: F821
    feedback: Mapped[list[ProjectFeedback]] = relationship(back_populates="review", cascade="all, delete-orphan")


class ProjectFeedback(UUIDPkMixin, TimestampMixin, Base):
    __tablename__ = "project_feedback"
    __table_args__ = (
        UniqueConstraint("review_id", "project_id", "reviewer_id", name="uq_project_feedback_reviewer"),
        CheckConstraint("status IN ('pending', 'submitted')", name="ck_project_feedback_status"),
    )

    review_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("performance_reviews.id", ondelete="CASCADE"), index=True
    )
    project_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("projects.id", ondelete="CASCADE"))
    reviewer_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("employees.id"), index=True)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="pending", server_default="pending")
    rating: Mapped[int | None] = mapped_column(Integer)
    contribution: Mapped[str | None] = mapped_column(Text)
    collaboration: Mapped[str | None] = mapped_column(Text)
    submitted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    review: Mapped[PerformanceReview] = relationship(back_populates="feedback")
    project: Mapped[Project] = relationship("Project")  # noqa: F821
    reviewer: Mapped[Employee] = relationship("Employee")  # noqa: F821

File: frontend/src/api/performance.ts
import { apiClient } from "./client";

export interface PerformanceCycle {
  id: string; name: string; description?: string | null; start_date: string; end_date: string;
  goal_due_date: string; self_review_due_date: string; manager_review_due_date: string;
  status: string; created_by_name?: string | null; approved_by_name?: string | null;
}
export interface PerformanceGoal {
  id: string; title: string; description?: string | null; category: string; measurement: string;
  weight: number; target_date: string; progress: number; evidence?: string | null; status: string;
  manager_comment?: string | null; version: number;
}
export interface ProjectFeedback {
  id: string; review_id: string; project_id: string; project_name: string; employee_name: string;
  status: string; rating?: number | null; contribution?: string | null; collaboration?: string | null;
}
export interface PerformanceReview {
  id: string; cycle: PerformanceCycle; employee_id: string; employee_name: string;
  manager_id?: string | null; manager_name: string; status: string; goals: PerformanceGoal[];
  self_summary?: string | null; self_rating?: number | null; manager_summary?: string | null;
  manager_rating?: number | null; calibration_comment?: string | null; calibrated_rating?: number | null;
  final_rating?: number | null; calibrated_by_name?: string | null; project_feedback: ProjectFeedback[];
  published_at?: string | null; acknowledged_at?: string | null; employee_comment?: string | null;
}
export interface PerformanceDashboard {
  cycles: PerformanceCycle[]; my_reviews: PerformanceReview[]; team_reviews: PerformanceReview[];
  feedback_requests: ProjectFeedback[]; can_manage_cycles: boolean; can_approve_cycles: boolean;
}

export type CyclePayload = Omit<PerformanceCycle, "id" | "status" | "created_by_name" | "approved_by_name">;
export type GoalPayload = Omit<PerformanceGoal, "id" | "status" | "manager_comment" | "version">;

export const performanceApi = {
  dashboard: () => apiClient.get<PerformanceDashboard>("/performance/dashboard").then(r => r.data),
  review: (id: string) => apiClient.get<PerformanceReview>(`/performance/reviews/${id}`).then(r => r.data),
  createCycle: (payload: CyclePayload) => apiClient.post("/performance/cycles", payload).then(r => r.data),
  updateCycle: (id: string, payload: CyclePayload) => apiClient.put(`/performance/cycles/${id}`, payload).then(r => r.data),
  submitCycle: (id: string) => apiClient.post(`/performance/cycles/${id}/submit`).then(r => r.data),
  decideCycle: (id: string, decision: "approve" | "changes_requested", comment?: string) => apiClient.post(`/performance/cycles/${id}/decision`, {decision, comment}).then(r => r.data),
  createGoal: (reviewId: string, payload: GoalPayload) => apiClient.post(`/performance/reviews/${reviewId}/goals`, payload).then(r => r.data),
  updateGoal: (goalId: string, payload: GoalPayload) => apiClient.put(`/performance/goals/${goalId}`, payload).then(r => r.data),
  submitGoals: (reviewId: string) => apiClient.post(`/performance/reviews/${reviewId}/goals/submit`).then(r => r.data),
  decideGoals: (reviewId: string, decision: "approve" | "changes_requested", comment?: string) => apiClient.post(`/performance/reviews/${reviewId}/goals/decision`, {decision, comment}).then(r => r.data),
  selfReview: (reviewId: string, summary: string, rating: number) => apiClient.post(`/performance/reviews/${reviewId}/self-review`, {summary, rating}).then(r => r.data),
  managerReview: (reviewId: string, summary: string, rating: number) => apiClient.post(`/performance/reviews/${reviewId}/manager-review`, {summary, rating}).then(r => r.data),
  feedback: (id: string, rating: number, contribution: string, collaboration?: string) => apiClient.post(`/performance/feedback/${id}`, {rating, contribution, collaboration}).then(r => r.data),
  calibrate: (reviewId: string, rating: number, comment: string) => apiClient.post(`/performance/reviews/${reviewId}/calibrate`, {rating, comment}).then(r => r.data),
  publish: (reviewId: string) => apiClient.post(`/performance/reviews/${reviewId}/publish`).then(r => r.data),
  acknowledge: (reviewId: string, comment?: string) => apiClient.post(`/performance/reviews/${reviewId}/acknowledge`, {comment}).then(r => r.data),
};
